Support seq(tensor), implement 2 sequence ops that use the new type.

pranavsharma opened the pull request:
**Description**: Support seq(tensor), implement 2 sequence ops that uses the new type.

**Motivation and Context**
The current implementation of seq(tensor) type is inadequate in that it doesn't account for the shape of the contained tensors. A bunch of sequence ops were added to opset11 that uses this type.

pranavsharma commented on onnxruntime/core/providers/cpu/sequence/sequence_ops.cc:
```diff
@@ -0,0 +1,147 @@
+    case ONNX_NAMESPACE::TensorProto_DataType_STRING:
+    case ONNX_NAMESPACE::TensorProto_DataType_FLOAT16:
+    case ONNX_NAMESPACE::TensorProto_DataType_BFLOAT16:
+      ORT_THROW("String, float16 and bloat16 Not implemented");
```
Should we support these in the first implementation?

pranavsharma replied:
fixed

skottmckay commented on include/onnxruntime/core/framework/tensor.h:
```diff
@@ -74,6 +76,7 @@ class Tensor final {
   Tensor(MLDataType p_type, const TensorShape& shape, std::shared_ptr<IAllocator> allocator, int64_t offset = 0);
 
+  void InitTensor(MLDataType p_type, const TensorShape& shape, std::shared_ptr<IAllocator> allocator, int64_t offset = 0);
   ~Tensor();
```
Constructing a Tensor and using std::move semantics to update an existing instance would be a bit cleaner.

pranavsharma replied:
Sure, this would work too. Why would exposing an Init method be any less clean though? You've a default constructed object and an Init method to initialize it.

skottmckay replied:
The existing code has move assignment. 

I'd argue that the semantics are clearer to go via a constructor and move assignment. Constructor always starts with a clean slate so it's clear what it needs to handle. InitTensor could be called at any point in time, and multiple times, so potentially you start adding logic (in theory) to take care of cleaning up any existing state to get back to a clean state in order to do the init. e.g. should the init be running the same logic as the destructor to cleanup any buffer that exists at the time?

---
In reply to: [331361908](https://github.com/microsoft/onnxruntime/pull/1983#discussion_r331361908) [](ancestors = 331361908)

pranavsharma replied:
Ok.

skottmckay commented on onnxruntime/core/framework/data_types.cc:
```diff
@@ -897,7 +936,7 @@ MLDataType DataTypeImpl::TypeFromProto(const ONNX_NAMESPACE::TypeProto& proto) {
   return type;
-}
+}  // namespace onnxruntime
 
```
Accidental change?

pranavsharma replied:
fixed

skottmckay commented on onnxruntime/core/framework/data_types.cc:
```diff
@@ -619,10 +638,7 @@ void RegisterAllProtos(const std::function<void(MLDataType)>& reg_fn) {
-  REGISTER_ONNX_PROTO(VectorInt64, reg_fn);
-  REGISTER_ONNX_PROTO(VectorDouble, reg_fn);
+  REGISTER_SEQ_TENSOR_PROTO(float, reg_fn);
 
```
What's the reason for removing these?

pranavsharma replied:
Those were the original registrations that represented seq(tensor) as vector<float>, etc. However, this reminded me to register other types as well here. Will do.

skottmckay commented on onnxruntime/core/providers/cpu/cpu_execution_provider.cc:
```diff
@@ -1,3 +1,4 @@
+
 // Copyright (c) Microsoft Corporation. All rights reserved.
```
nit: rogue newline

pranavsharma replied:
fixed

skottmckay commented on onnxruntime/core/providers/cpu/sequence/sequence_ops.cc:
```diff
@@ -0,0 +1,147 @@
+    SequenceAt);
+
+static void GetSeqIdx(const Tensor& idx_tensor, int64_t& seq_idx) {
+  auto idx_tensor_dtype = utils::GetTensorProtoType(idx_tensor);
```
Would returning seq_idx be a little more natural than passing it as a non-const argument?

pranavsharma replied:
I did it this way so that I could throw an exception for an unsupported data type.

pranavsharma replied:
sure.

skottmckay commented on onnxruntime/core/providers/cpu/sequence/sequence_ops.cc:
```diff
@@ -0,0 +1,147 @@
+
+template <typename T>
+static void CopyTensor(const Tensor& indexed_tensor, Tensor& output_tensor) {
+  const auto* input_data = indexed_tensor.template Data<T>();
```
Can you use CopyCpuTensor from providers\cpu\tensor\utils.h?

pranavsharma replied:
sure

skottmckay replied:
Can this be deleted now?

---
In reply to: [331652371](https://github.com/microsoft/onnxruntime/pull/1983#discussion_r331652371) [](ancestors = 331652371)

skottmckay commented on onnxruntime/core/providers/cpu/sequence/sequence_ops.cc:
```diff
@@ -0,0 +1,147 @@
+  memcpy(output_data, input_data, indexed_tensor.SizeInBytes());
+}
+
+static void FillOutputTensor(const Tensor& indexed_tensor, Tensor& output_tensor) {
```
>FillOutputTensor [](start = 12, length = 16)

nit: 'Fill' somewhat implies multiple copies. Maybe just CopyToOutput

pranavsharma replied:
replaced with CopyCpuTensor

skottmckay commented on onnxruntime/test/providers/cpu/sequence/sequence_ops_test.cc:
```diff
@@ -0,0 +1,78 @@
+TEST(SequenceOpsTest, SequenceLengthPositive) {
+  OpTester test("SequenceLength", 11);
+  SeqTensors<float> input;
+  using PairType = std::pair<std::vector<int64_t>, std::vector<float>>;
```
>PairType [](start = 8, length = 8)

Maybe ShapeAndTensor would be clearer?

pranavsharma replied:
changed.

skottmckay commented on onnxruntime/test/providers/provider_test_utils.h:
```diff
@@ -26,6 +26,11 @@ namespace onnxruntime {
+template <typename T>
+struct SeqTensors {
+  std::vector<std::pair<std::vector<int64_t>, std::vector<T>>> tensors;
+};
```
A comment explaing the first item in the pair is the shape would be helpful

pranavsharma replied:
done

pranavsharma commented:
The recent commits make 2 updates
- I changed the data type from vector<Tensor> to TensorSeq (see comments and explanation in data_types.h)
- I've also added support in the python API for inputs. Support for outputs will come in a follow up PR so that this PR doesn't block the implementation of the remaining sequence ops.

pranavsharma commented:
@skottmckay all the mandatory builds have passed and this is ready to to be reviewed again. Thanks!

skottmckay commented on onnxruntime/core/session/onnxruntime_c_api.cc:
```diff
@@ -773,32 +793,61 @@ OrtStatus* PopulateTensorWithData<std::string>(OrtValue* oval, const std::string
+    st = OrtGetValueImplSeqOfTensorsHelper<std::string>(allocator, one_tensor, out);
+  } else {
+    st = OrtApis::CreateStatus(ORT_FAIL, "Only sequences that contain float tensors are supported.");
+  }
```
Is this error message correct?

pranavsharma replied:
Ah, crap! Thanks for identifying. I'm writing the code for a few other Sequence ops and will fix it as part of the follow up PR.

skottmckay commented on onnxruntime/core/session/onnxruntime_c_api.cc:
```diff
@@ -897,44 +946,106 @@ ORT_API_STATUS_IMPL(OrtApis::GetValue, const OrtValue* value, int index, OrtAllo
+
+  // use the data type of the first tensor as the data type of the seq
+  seq_ptr->dtype = reinterpret_cast<const OrtValue*>(in[0])->Get<Tensor>().DataType();
+
```
Should TensorSeq have a ctor that requires the data type to be set?

pranavsharma replied:
Yeah, let me check if there're other problems in doing this. Will address in follow up PR.

skottmckay commented on onnxruntime/core/session/onnxruntime_c_api.cc:
```diff
@@ -897,44 +946,106 @@ ORT_API_STATUS_IMPL(OrtApis::GetValue, const OrtValue* value, int index, OrtAllo
-    auto data = tensor.Data<T>();
-    if (!data) {
-      return OrtApis::CreateStatus(ORT_FAIL, "Encountered nullptr.");
+    auto& one_tensor = reinterpret_cast<const OrtValue*>(in[idx])->Get<Tensor>();
```
>reinterpret_cast<const OrtValue*> [](start = 23, length = 33)

Why do we need the reinterpret_cast?

pranavsharma replied:
May be we don't. Probably a copy-paste error. Will address in follow up PR.

skottmckay commented on onnxruntime/core/framework/TensorSeq.h:
```diff
@@ -0,0 +1,42 @@
+  // TODO: optimization opportunity - if all tensors in the seq are scalars, we can potentially represent them
+  // as vector<primitive type>
+  std::vector<Tensor> tensors;
+};
```
What frees the Tensor instances?
